compaction.js

```diff
@@ -0,0 +1,371 @@
+// SPDX-FileCopyrightText: 2022 Anders Rune Jensen
+//
+// SPDX-License-Identifier: LGPL-3.0-only
+
+const RAF = require('polyraf')
+const fs = require('fs')
+const Obv = require('obz')
+const push = require('push-stream')
+const mutexify = require('mutexify')
+const debug = require('debug')('async-append-only-log')
+const Record = require('./record')
+
+function getStateFilename(logFilename) {
+  return logFilename + '.compaction'
+}
+
+function stateFileExists(logFilename) {
+  return fs.existsSync(getStateFilename(logFilename))
+}
+
+/**
+ * This file has state describing the continuation of the compaction algorithm.
+ *
+ * - bytes 0..3: UInt32LE for the blockIndex to-be-compacted
+ * - bytes 4..7: UInt32LE for the 1st unshifted record's offset
+ * - bytes 8..(8+blockSize-1): blockBuf containing the 1st unshifted record
```

Comment:
how do you keep track of the following case:

you have a block where the last record is deleted, so you need to "move" the first record from the next block

It seems like there is a state where you have written the block with the hole, but how do you keep track (in case of a crash), that the first record (or several) needs to be removed from the next block.

Reply:
I think there's a test case for that, should be in `test/compaction.js` called "recovers from crash just after persisting state". See also "recovers from crash just after persisting state".

> It seems like there is a state where you have written the block with the hole, but how do you keep track (in case of a crash), that the first record (or several) needs to be removed from the next block.

The *second* record from the next block would be the "unshifted record" (because the first record has already been "shifted"), and this is saved together with the next block as the blockBuf in the state file *before* running the compaction algorithm.

Reply:
Ohhh, this is pretty smart. The offset can be in another buffer. I didn't see this at all when reading this line but see now that it is documented further down.